lineapy/transformer/linea_transformer.py

```diff
@@ -0,0 +1,171 @@
+import ast
+import logging
+from pathlib import Path
+from types import ModuleType
+from typing import Optional, Union
+
+from lineapy.api.api import _artifact_store, _get, _save, healthcheck
+from lineapy.data.types import (
+    LineaCallNode,
+    Node,
+    PositionalArgument,
+    SourceCode,
+    SourceCodeLocation,
+    SourceLocation,
+)
+from lineapy.exceptions.db_exceptions import ArtifactSaveException
+from lineapy.instrumentation.tracer import Tracer
+from lineapy.utils.utils import get_new_id
+
+logger = logging.getLogger(__name__)
+
+
+class LineaTransformer(ast.NodeTransformer):
+    def __init__(
+        self,
+        code: str,
+        location: SourceCodeLocation,
+        tracer: Tracer,
+    ):
+        self.source_code = SourceCode(
+            id=get_new_id(), code=code, location=location
+        )
+        # tracer.db.write_source_code(self.source_code)
+        self.tracer = tracer
+        # Set __file__ to the pathname of the file
+        if isinstance(location, Path):
+            tracer.executor.module_file = str(location)
+        # The result of the last line, a node if it was an expression,
+        # None if it was a statement. Used by ipython to grab the last value
+        self.last_statement_result: Optional[Node] = None
+
+    def lgeneric_visit(self, node: ast.AST):
+        return node
```

Comment:
Curious why we had to change from the `ast` lib's `NodeVisitor`'s `generic_visit` to our own generic visit?